### PR TITLE
Use webkms-context.

### PR DESCRIPTION
**Pull request by JSAssassin**

### CODE REVIEW

**Review by mattcollier — APPROVED** [review]
LGTM, needs changelog entry for what I believe is another major breaking release.

**Review by dmitrizagidulin — APPROVED** [review]
:+1: 
(and yeah, agreed, breaking change.)

File: lib/schemas/operationContext.js
Comment (mattcollier) [inline lib/schemas/operationContext.js:9]: So here is a comment indicating that, at least in the case of a `GenerateKeyOperation` we need to have  the context for the key such as `Ed25519VerificationKey2020` `https://w3id.org/security/suites/ed25519-2020/v1`.

https://github.com/digitalbazaar/webkms-client/pull/24#discussion_r614308941

I have made this the case which is running afoul of this validation.

Before I proceed, I want to ensure that what is needed here is that we need to allow for:
- Array index 0 is always `webkms-context`
- Array index 1 is one of the context for one of the supported key types here: https://github.com/digitalbazaar/webkms-client/blob/56230b10d5de668d44dbdee1d6004d18c8dc445b/cryptoLd.js#L15-L19

Is all of this correct?

@dlongley @dmitrizagidulin

Comment (dlongley) [reply lib/schemas/operationContext.js:9]: @mattcollier, I looked at the schemas and it looks like we only need the key suite context when we're generating a key. Otherwise we don't. I think we should not use the `operationContext` in the `GenerateKeyOperation` schema and special case it there ... or take some other similar approach.

Comment (dmitrizagidulin) [reply lib/schemas/operationContext.js:9]: I don't think the context array is order-preserving - right @dlongley? I vaguely recall we tried to fix the order for Verifiable Credentials, but that had more to do with systems that consumed the VCs as regular JSON.

**Review by dmitrizagidulin — APPROVED** [review]
:+1: 
lgtm

**Comment by mattcollier** [comment]
Before this latest round of schema edits, the schema was not working with the code here, so the new schemas should be tested with https://github.com/digitalbazaar/bedrock-kms-http/tree/support-ed25519-2020 before merge.

File: lib/schemas/operationContext.js
Comment (dlongley) [inline lib/schemas/operationContext.js:36]: I don't think the above is what we want. We should not be accepting unnecessary suite contexts on non-GenerateKeyOperations. Only the gen key op needs a context with an array that includes the suite for the key that is being generated. The other ops should just use a string value for the context and it must be the WebKMS context.

Comment (dmitrizagidulin) [reply lib/schemas/operationContext.js:36]: I see. so, make a separate schema just for the generate key operation (and use this array). And all the other operations just have the one context.

**Review by dlongley — CHANGES_REQUESTED** [review]
Every op except the key gen one should require the `@context` value to be a string using the webkms/v1 context. The gen key one should require `@context` value to be an array with the first entry as the webkms/v1 context -- and the second entry as the context that goes along with the key that is being generated. I'm not sure we want to do this second check via JSON schema, but if we do, it makes more sense to me to put the whole thing in the gen key operation schema since we'll need to match up key types with the appropriate context.

File: lib/schemas/operationContext.js
Comment (dlongley) [reply lib/schemas/operationContext.js:36]: @dmitrizagidulin Yeah, that's one option -- except that we could still wind up with some broken things. We don't want to accept an operation with a context of X for a key type of Y. So this array as it stands is just accepting *any* of the possible context combinations without linking it to the actual data (the `type` used in the data).

**Comment by dmitrizagidulin** [comment]
@mattcollier 
> Before this latest round of schema edits, the schema was not working with the code here, so the new schemas should be tested with https://github.com/digitalbazaar/bedrock-kms-http/tree/support-ed25519-2020 before merge.

Yeah, I just ran the tests over on that bedrock-kms-http PR with this branch, and some of the tests failed, so we need to debug further.

**Comment by JSAssassin** [comment]
> @mattcollier
> 
> > Before this latest round of schema edits, the schema was not working with the code here, so the new schemas should be tested with https://github.com/digitalbazaar/bedrock-kms-http/tree/support-ed25519-2020 before merge.
> 
> Yeah, I just ran the tests over on that bedrock-kms-http PR with this branch, and some of the tests failed, so we need to debug further.

@dmitrizagidulin @mattcollier Those failing tests(all with permission denied error) seems to be because of some validation failure in @digitalbazaar/zcapld. here's a gist of the error stack. https://gist.github.com/JSAssassin/8748078998d28dd410f8c691b9ef2f58

**Comment by mattcollier** [comment]
> @dmitrizagidulin @mattcollier Those failing tests(all with permission denied error) seems to be because of some validation failure in @digitalbazaar/zcapld. here's a gist of the error stack. https://gist.github.com/JSAssassin/8748078998d28dd410f8c691b9ef2f58

@JSAssassin good, this is progress since we're now getting past the schema validation in webkms-switch.  The issue at this point appears to be here :

`fetchInSecurityContext`
https://github.com/digitalbazaar/zcapld/blob/main/lib/CapabilityInvocation.js#L113-L114
defined here;
https://github.com/digitalbazaar/zcapld/blob/main/lib/utils.js#L140

This capability:
```
{
  '@context': 'https://w3id.org/webkms/v1',
  id: 'https://localhost:18443/kms/keystores/z1A7uKtayX8MQVmnQtUzAnDhm/keys/z19xPrwH3X9UVg9n8jeH4Hkf1',
  type: 'Ed25519VerificationKey2020',
  kmsModule: 'ssm-v1',
  publicKeyMultibase: 'z3fQNuwm4JhdaSVHsA61G8Rw8dFP4UbNZGN2ArDZ1Mm1M',
  controller: 'did:key:z6MkhRjrvpEAxh8oSWsPgMdP28xvrotsDBCvcwTtbUuUEp38'
}
```

is being transformed to:
```
{
  '@context': 'https://w3id.org/security/v2',
  id: 'https://localhost:18443/kms/keystores/z1A7uKtayX8MQVmnQtUzAnDhm/keys/z19xPrwH3X9UVg9n8jeH4Hkf1',
  type: 'Ed25519VerificationKey2020',
  kmsModule: 'ssm-v1'
}
```

As you can see, terms are being dropped in this process.  Notably `controller`.

I suggest making it so that if the context on the capability is `https://w3id.org/webkms/v1` that you skip compaction as currently the case with some other set of contexts.

File: lib/schemas/generateKeyOpContext.js
Comment (mattcollier) [inline lib/schemas/generateKeyOpContext.js:5]:
```suggestion
'use strict';

const {CONTEXT_URL: WEBKMS_CONTEXT_URL} = require('webkms-context');
```

**Review by mattcollier — CHANGES_REQUESTED** [review]

File: lib/schemas/operationContext.js
Comment (JSAssassin) [reply lib/schemas/operationContext.js]: addressed here 4d57829

File: lib/schemas/generateKeyOpContext.js
Comment (JSAssassin) [reply lib/schemas/generateKeyOpContext.js:16]: updated here 4d57829

File: lib/schemas/generateKeyOpContext.js
Comment (JSAssassin) [reply lib/schemas/generateKeyOpContext.js]: Fixed 4d57829

**Review by dmitrizagidulin — APPROVED** [review]
:+1:

File: lib/index.js
Comment (dlongley) [inline lib/index.js:114]: We also need to check that the suite context is in the message, right? The json-schema we have will make sure at least *one* of the possible key suite contexts we support will be present, however, it won't check to make sure that the one we need that *matches* `invocationTarget.type` is present. For example, someone may pass a key gen op requesting to create an ed25519 key -- but use the aes keywrap context. We should also have a test to make sure this won't be accepted.

**Review by dlongley — APPROVED** [review]
I'm approving conditionally based on my comments. There's just a bit more to do here but once finished I don't expect it to need another review from me.

File: lib/index.js
Comment (JSAssassin) [reply lib/index.js:114]: Done here ddbb378

**Review by mattcollier — APPROVED** [review]
TY.

**Pull request merged**
